mono/mini/mini-arm64.c

```diff
@@ -2670,29 +2675,66 @@ mono_arch_emit_setret (MonoCompile *cfg, MonoMethod *method, MonoInst *val)
 	}
 }
 
+static const gboolean debug_tailcall = FALSE;
+
+static gboolean
+is_supported_tailcall_helper (gboolean value, const char *svalue)
+{
+	if (!value && debug_tailcall)
+		g_print ("%s %s\n", __func__, svalue);
+	return value;
+}
+
+#define IS_SUPPORTED_TAILCALL(x) (is_supported_tailcall_helper((x), #x))
+
+static gboolean
+tailcall_return_storage_supported (ArgStorage storage)
+{
+	switch (storage) {
+	case ArgInIReg:
+	case ArgInFReg:
+	case ArgNone:
+		return TRUE;
+
+	case ArgHFA:
+	case ArgInFRegR4: // conversion in emit_move_return_value missed
```

Comment:
This is overly pessimistic I now understand and can move up into the supported area. In particular, float32 does not change the ABI, and there is the check that caller and callee have same return storage.